cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/connection/ConnectionStore.java

```diff
@@ -0,0 +1,164 @@
+   * @throws ConnectionNotFoundException if the connection is not found
+   */
+  public Connection getConnection(ConnectionId connectionId) throws ConnectionNotFoundException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TABLE_ID);
+      return getConnectionInternal(table, connectionId, true).get();
+    }, ConnectionNotFoundException.class);
+  }
+
+  /**
+   * Get all the connections in the given namespace
+   *
+   * @param namespace the namespace to look up
+   * @return the list of connections in this namespace
+   */
+  public List<Connection> listConnections(NamespaceSummary namespace) {
+    return TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TABLE_ID);
+      Range range = Range.singleton(getNamespaceKeys(namespace));
+      List<Connection> connections = new ArrayList<>();
+      try (CloseableIterator<StructuredRow> rowIter = table.scan(range, Integer.MAX_VALUE)) {
+        rowIter.forEachRemaining(
+          structuredRow -> connections.add(GSON.fromJson(structuredRow.getString(CONNECTION_DATA_FIELD),
+                                                         Connection.class)));
+      }
+      return connections;
+    });
+  }
+
+  /**
+   * Save the connection in the store.
+   *
+   * @param connectionId the connection id
+   * @param connection the connection information
+   */
+  public void saveConnection(ConnectionId connectionId, Connection connection) {
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TABLE_ID);
+      Optional<Connection> oldConnection  = getConnectionInternal(table, connectionId, false);
+      Connection newConnection = connection;
+      if (oldConnection.isPresent()) {
+        newConnection = new Connection(connection.getName(), connection.getConnectionType(),
+                                       connection.getDescription(), connection.isPreConfigured(),
+                                       oldConnection.get().getCreatedTimeMillis(), connection.getPlugin());
+      }
+      Collection<Field<?>> fields = getConnectionKeys(connectionId);
+      fields.add(Fields.stringField(CONNECTION_DATA_FIELD, GSON.toJson(newConnection)));
+      table.upsert(fields);
+    });
+  }
+
+  /**
+   * Delete the given connection
+   *
+   * @param connectionId the connection id to delete
+   * @throws ConnectionNotFoundException if the connection is not found
+   */
+  public void deleteConnection(ConnectionId connectionId) throws ConnectionNotFoundException {
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TABLE_ID);
+      getConnectionInternal(table, connectionId, true);
+      table.delete(getConnectionKeys(connectionId));
+    }, ConnectionNotFoundException.class);
+  }
+
+  // use this method to fetch connection outside a transaction
```

Comment:
what does it mean to fetch outside a transaction? Doesn't everything need to be within a transaction?

Reply:
This is just for internal use so that when delete, create tries to fetch the existing connection, it will happen in a single transaction(saw draft store it happens in two transactions), I have updated the comment to be more clear on this.